File: Codility/lessons/PermMissingElem.js
/**
 * Lesson 3. Time Complexity
 */
function solution(A) {
    const N = A.length;
    const check = new Array(N + 2).fill(0);
    for (let num of A) {
        check[num] = 1;
    }
    for (let i = 1; i <= N + 1; i++) {
        if (check[i] == 0) {
            return i;
        }
    }
}

File: Codility/src/lesson3/PermMissingElem.java
package lesson3;

public class PermMissingElem {
    public int solution(int[] A) {
        int N = A.length;
        long expectedSum = (long) (N + 1) * (N + 2) / 2;
        long missingSum = 0L;
        for (int num : A) {
            missingSum += num;
        }
        return (int) (expectedSum - missingSum);
    }

    public int solution_array(int[] A) {
        int N = A.length;
        boolean[] check = new boolean[N + 2];
        for (int num : A) {
            check[num] = true;
        }

        int missing = -1;
        for (int num = 1; num <= N + 1; num++) {
            if (!check[num]) {
                missing = num;
                break;
            }
        }
        return missing;
    }
}
